components/datetime/src/lib.rs

```diff
@@ -204,35 +204,35 @@ mod tests {
 
     #[test]
     fn check_sizes() {
-        check_size_of!(5800 | 4592, DateFormatter);
-        check_size_of!(6792 | 5456, DateTimeFormatter);
-        check_size_of!(7904 | 6480, ZonedDateTimeFormatter);
-        check_size_of!(1496 | 1320, TimeFormatter);
+        check_size_of!(5800 | 4632, DateFormatter);
```

Comment:
I bet the compiler was previously able to "merge" the `T` with whatever other fields were present on the struct, but now with the MaybeUninit it is forced to keep the payload as its own chunk of memory. A bit sad.

Reply:
yeah it only makes sense if there's an enum wrapping it, and the only enum I see is the thing that swaps between yokes and staticrefs, and I'm not quite sure how that could have a size impact.

Reply:
It looks like if the payload of two variants have a common part, and if one of payloads separately has enough of a niche for Rust to be able to stuff the `enum`'s discriminant inside, then it does so, effectively reducing the size of the whole `enum`. In which case the niche removal of `MaybeUninit` does effectively impact the overall size.

```rs
use ::core::mem::{MaybeUninit as MU, size_of};

enum Enum<T> {
    A(T, u8),
    B(   u8),
}

assert_eq!(2, size_of::<Enum<    bool  >>());
assert_eq!(3, size_of::<Enum< MU<bool> >>());
```

 - [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=744b4f3ba877395f508ed42d0a21b19c)

Reply:
aha! thanks for investigating!

Though the `DataPayload` enum here [doesn't have variants with common parts before it gets to the yoke](https://unicode-org.github.io/icu4x/docs/src/icu_provider/response.rs.html#74-78)